# IS 362 Assignment 12 
##  Preprocessing Data for scikit-learn
## Taylor Clark Jones
Very often, we’re tasked with taking data in one form and transforming it for easier downstream analysis. In this
assignment, you’ll use what you’ve learned in the course to prepare data for predictive analysis in Project 4.

We'll first start this project by loading in the dataset and adding meaningful names for each column,

In [66]:
import pandas as pd
import matplotlib.pyplot as plt

In [67]:
#column names taken from attribute list feqatured on: https://archive.ics.uci.edu/ml/datasets/Mushroom
columns =  ['target', 'cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
           'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color', 'stalk-shape',
           'stalk-root', 'stalk-surface-above-ring', 'stack-surface-below-ring',
           'stalk-color-above-ring', 'stalk-color-below-ring', 'veil-type', 'veil-color',
           'ring-number', 'ring-type', 'spore-print-color', 'population', 'habitat']
df = pd.read_csv('file:///C:/Users/HP/Downloads/agaricus-lepiota.data', names=columns)

In [68]:
df.head()



target cap-shape cap-surface cap-color bruises odor gill-attachment  \
0      p         x           s         n       t    p               f   
1      e         x           s         y       t    a               f   
2      e         b           s         w       t    l               f   
3      p         x           y         w       t    p               f   
4      e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stack-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

####  Create a pandas DataFrame with a subset of the columns in the dataset. You should include the column that indicates edible or poisonous, the column that includes odor, and at least one other column of your choosing.


In [69]:
df_subset = df[['target', 'odor', 'gill-color']]

#### Replace the codes used in the data with numeric values—for example, in the first “target” column, “e” might become 0 and “p” might become 1. This is because your downstream processing in Project 4 using scikit-learn requires that values be stored as numerics.

In [70]:
df_subset['target'].replace({'p': 1, 'e': 0}, inplace=True)


C:\Users\HP\AppData\Local\Temp\ipykernel_84472\2888384435.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subset['target'].replace({'p': 1, 'e': 0}, inplace=True)


In [71]:
# check to ensure values have changed
df_subset.head()

target odor gill-color
0       1    p          k
1       0    a          k
2       0    l          n
3       1    p          n
4       0    n          k

#### exploratory data analysis: I will show the distribution of data for each of the columns I've selected, and show scatterplots for edible/poisonous vs. odor as well as the other column that you selected.

In [72]:
df_subset['target'].value_counts(normalize=True)

0    0.517971
1    0.482029
Name: target, dtype: float64

For the target column the variable '0' represents edible, while the variable '1' represents poisonous.
According to the decimals there are more edible mushrooms with a decimal of '0.517971' that can be converted and rounded to 52%.

While poisonous mushrooms, with a decimal of '0.482029', can br converted and rounded t0 48% of mushrooms.

In [73]:
# for reference the letters displayed under odor is as follows: almond=a,anise=l,creosote=c,fishy=y,foul=f, musty=m,none=n,pungent=p,spicy=s
df_subset['odor'].value_counts(normalize=True)

n    0.434269
f    0.265879
y    0.070901
s    0.070901
a    0.049237
l    0.049237
p    0.031512
c    0.023634
m    0.004431
Name: odor, dtype: float64

Most mushrooms have no order, with 'n' (non) having a decimal of '0.434269' which converts to 43% of mushrooms.

With '0.265879' (or 27%) of mushrooms being assigned to 'f', a fishy order.

In [74]:
# for reference the letter displayed under gill-color is as follows: black=k,brown=n,buff=b,chocolate=h,gray=g, green=r,orange=o,pink=p,purple=u,red=e, white=w,yellow=y
df_subset['gill-color'].value_counts(normalize=True)

b    0.212703
p    0.183653
w    0.147957
n    0.129000
g    0.092565
h    0.090103
u    0.060561
k    0.050222
e    0.011817
y    0.010586
o    0.007878
r    0.002954
Name: gill-color, dtype: float64

The most common gill color is 'b' which stands for buff with a decimal of '0.212703' (or 21%). With 'p', which stands for pink, having a decimal of '0.183653' (18%) and 'w', or white, having a decimal of '9.147957' (15%).

The gill color of a mushroom is a prominent feature of a mushroom that allows people to see if a mushroom is poisonous or not. According to sciencing.com poisonous mushrooms also have "with a ring around the stem and those with a volva."
website: https://sciencing.com/identify-poisonous-mushrooms-2057768.html

### edible/poisonous mushrooms Vs. mushroom odor

In the same science.com article, they state "Poisonous mushrooms often have an unpleasant, acrid smell, while benign ones smell refreshingly mushroomlike."

We will cross-check the smell of a mushroom (odor) with the edibility of a mushroom (target).

In [75]:
#As previously stated variable 'o' is edible while '1' is poisonous.
df_subset.groupby(['odor', 'target']).count()

gill-color
odor target            
a    0              400
c    1              192
f    1             2160
l    0              400
m    1               36
n    0             3408
     1              120
p    1              256
s    1              576
y    1              576

Supporting the science article, smells 'fishy', 'spicy' , foul' , and other unpleasant smells correspond with the variable '1' marking them as poisonous.
While smells like 'almond' , 'anise' and 'none' are safe to consume.

I mentioned gill color being a indicator of a poisonous mushroom earlier within this notebook. We will now cross examine both columns (gill-color and target0 to verify.

In [76]:
df_subset.groupby(['gill-color', 'target']).count()

odor
gill-color target      
b          1       1728
e          0         96
g          0        248
           1        504
h          0        204
           1        528
k          0        344
           1         64
n          0        936
           1        112
o          0         64
p          0        852
           1        640
r          1         24
u          0        444
           1         48
w          0        956
           1        246
y          0         64
           1         22

As you can see the color of a mushroom can be an indicator of edibility but colors such as 'w', 'y' , 'p' , and 'g' have both poisonous and edible mushrooms matching their color description.
However mushrooms with a gill-color of 'b' or 'buff' are all poisonous.  